File: cpp/tests/unit-tests/AceValidatorTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

#include "joynr/PrivateCopyAssign.h"
#include <gtest/gtest.h>
#include "cluster-controller/access-control/AceValidator.h"
using namespace ::testing;
using namespace joynr;
using namespace joynr::infrastructure::DacTypes;

class AceValidatorTest : public ::testing::Test {
public:
    MasterAccessControlEntry masterAce;
    MasterAccessControlEntry mediatorAce;
    OwnerAccessControlEntry ownerAce;
public:
    AceValidatorTest()
    {
    }

    ~AceValidatorTest() = default;

    void SetUp(){
        std::vector<Permission::Enum> possiblePermissions;
        possiblePermissions.push_back(Permission::NO);
        possiblePermissions.push_back(Permission::ASK);
        std::vector<TrustLevel::Enum> possibleRequiredTrustLevels;
        possibleRequiredTrustLevels.push_back(TrustLevel::LOW);
        possibleRequiredTrustLevels.push_back(TrustLevel::MID);
        std::vector<TrustLevel::Enum> possibleRequiredAceChangeTrustLevels;
        possibleRequiredAceChangeTrustLevels.push_back(TrustLevel::MID);
        possibleRequiredAceChangeTrustLevels.push_back(TrustLevel::HIGH);
        masterAce = MasterAccessControlEntry(TEST_USER,
                                                                    std::string(),
                                                                    std::string(),
                                                                    TrustLevel::LOW,
                                                                    possibleRequiredTrustLevels,
                                                                    TrustLevel::LOW,
                                                                    possibleRequiredAceChangeTrustLevels,
                                                                    std::string(),
                                                                    Permission::NO,
                                                                    possiblePermissions);

        mediatorAce = MasterAccessControlEntry(TEST_USER,
                                               std::string(),
                                               std::string(),
                                               TrustLevel::LOW,
                                               possibleRequiredTrustLevels,
                                               TrustLevel::LOW,
                                               possibleRequiredAceChangeTrustLevels,
                                               std::string(),
                                               Permission::NO,
                                               possiblePermissions);

        ownerAce = OwnerAccessControlEntry(TEST_USER,
                                                                  std::string(),
                                                                  std::string(),
                                                                  TrustLevel::MID,
                                                                  TrustLevel::HIGH,
                                                                  std::string(),
                                                                  Permission::ASK);
    }

protected:
    static const std::string TEST_USER;
private:
    DISALLOW_COPY_AND_ASSIGN(AceValidatorTest);
};

const std::string AceValidatorTest::TEST_USER("testUser");

//----- Tests ------------------------------------------------------------------

TEST_F(AceValidatorTest, TestMediatorInvalidPossiblePermissions)
{
    std::vector<Permission::Enum> possiblePermissions;
    possiblePermissions.push_back(Permission::ASK);
    possiblePermissions.push_back(Permission::YES);
    mediatorAce.setPossibleConsumerPermissions(possiblePermissions);
    AceValidator validator(masterAce, mediatorAce, ownerAce);

    EXPECT_FALSE(validator.isMediatorValid());
}

TEST_F(AceValidatorTest, TestMediatorInvalidPossibleTrustLevels)
{
    std::vector<TrustLevel::Enum> possibleRequiredTrustLevels;
    possibleRequiredTrustLevels.push_back(TrustLevel::HIGH);
    possibleRequiredTrustLevels.push_back(TrustLevel::MID);
    mediatorAce.setPossibleRequiredTrustLevels(possibleRequiredTrustLevels);
    AceValidator validator(masterAce, mediatorAce, ownerAce);

    EXPECT_FALSE(validator.isMediatorValid());
}

TEST_F(AceValidatorTest, TestMediatorValid)
{
    AceValidator validator(masterAce, mediatorAce, ownerAce);

    EXPECT_TRUE(validator.isMediatorValid());
}

TEST_F(AceValidatorTest, TestOwnerValid)
{
    AceValidator validator(masterAce, mediatorAce, ownerAce);

    EXPECT_TRUE(validator.isOwnerValid());
}

TEST_F(AceValidatorTest, TestOwnerInvalid)
{
    ownerAce.setConsumerPermission(Permission::YES);
    AceValidator validator(masterAce, mediatorAce, ownerAce);
    EXPECT_FALSE(validator.isOwnerValid());
}

File: java/common/infrastructure-common/src/test/java/io/joynr/accesscontrol/AceValidatorTest.java
package io.joynr.accesscontrol;

/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import joynr.infrastructure.DacTypes.MasterAccessControlEntry;
import joynr.infrastructure.DacTypes.OwnerAccessControlEntry;
import joynr.infrastructure.DacTypes.Permission;
import joynr.infrastructure.DacTypes.TrustLevel;

import org.junit.Before;
import org.junit.Test;

public class AceValidatorTest {

    private MasterAccessControlEntry masterAce;
    private MasterAccessControlEntry mediatorAce;
    private OwnerAccessControlEntry ownerAce;

    @Before
    public void setup() {
        masterAce = new MasterAccessControlEntry(null,
                                                 null,
                                                 null,
                                                 TrustLevel.LOW,
                                                 new TrustLevel[]{ TrustLevel.MID, TrustLevel.LOW },
                                                 TrustLevel.LOW,
                                                 new TrustLevel[]{ TrustLevel.MID, TrustLevel.HIGH },
                                                 null,
                                                 Permission.NO,
                                                 new Permission[]{ Permission.ASK, Permission.NO });

        mediatorAce = new MasterAccessControlEntry(null,
                                                   null,
                                                   null,
                                                   TrustLevel.LOW,
                                                   new TrustLevel[]{ TrustLevel.MID, TrustLevel.LOW },
                                                   TrustLevel.LOW,
                                                   new TrustLevel[]{ TrustLevel.MID, TrustLevel.HIGH },
                                                   null,
                                                   Permission.NO,
                                                   new Permission[]{ Permission.ASK, Permission.NO });

        ownerAce = new OwnerAccessControlEntry(null, null, null, TrustLevel.MID, TrustLevel.HIGH, null, Permission.ASK);
    }

    @Test
    public void testMediatorInvalidPossiblePermissions() {
        mediatorAce.setPossibleConsumerPermissions(new Permission[]{ Permission.YES, Permission.ASK });
        AceValidator validator = new AceValidator(masterAce, mediatorAce, ownerAce);
        assertFalse("modified mediator is not valid", validator.isMediatorValid());
    }

    @Test
    public void testMediatorInvalidPossibleTrusLevels() {
        mediatorAce.setPossibleRequiredTrustLevels(new TrustLevel[]{ TrustLevel.HIGH, TrustLevel.MID });
        AceValidator validator = new AceValidator(masterAce, mediatorAce, ownerAce);
        assertFalse("modified mediator is not valid", validator.isMediatorValid());
    }

    @Test
    public void testMediatorValid() {
        AceValidator validator = new AceValidator(masterAce, mediatorAce, ownerAce);
        assertTrue("initial mediator is valid", validator.isMediatorValid());
    }

    @Test
    public void testOwnerValid() {
        AceValidator validator = new AceValidator(masterAce, mediatorAce, ownerAce);
        assertTrue("initial owner is valid", validator.isOwnerValid());
    }

    @Test
    public void testOwnerInvalid() {
        ownerAce.setConsumerPermission(Permission.YES);
        AceValidator validator = new AceValidator(masterAce, mediatorAce, ownerAce);
        assertFalse("modified owner is invalid", validator.isOwnerValid());
    }
}
